File: src/main/java/Manager/ImageManager.java
package Manager;

import java.io.File;
import java.net.MalformedURLException;
import java.util.HashMap;
import java.util.Map;

import Main.Config;
import javafx.scene.image.Image;

/**
 * the class is for cache icon
 * @author Administrator
 *
 */
public class ImageManager {

	public static final String ICON_TEXTURE = "icon_image";
	public static final String ICON_MOVIECLIP = "icon_movieclip";
	public static final String ICON_MUSIC = "icon_music";
	
	
	
	private static ImageManager instance;

	public static ImageManager getInstance()
	{
		if(instance==null)
		{
			instance=new ImageManager();
		}
		return instance;
		
	}
	
	private Map<String, Image> map=new HashMap<>();
	
	public ImageManager()
	{
		loadImage();
		
	}
	
	
	private void loadImage(){
		_loadImage(ICON_TEXTURE);
		_loadImage(ICON_MOVIECLIP);
		_loadImage(ICON_MUSIC);
	}

	private void _loadImage(String name){
		
		try {
			//System.out.println(new File(Config.IMAGE_DIRECTORY+name+".png").toURL().toString());
			map.put(name, new Image(new File(Config.IMAGE_DIRECTORY+name+".png").toURL().toString()));
		} catch (MalformedURLException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
	}


	public Image getIcon(String name)
	{
		return map.get(name);	
	}
}

File: src/main/java/Manager/ImageManager.kt
package Manager

import JavaFxPlus.KotlinEx.toFile
import Main.Config
import com.google.common.cache.CacheBuilder
import com.google.common.cache.CacheLoader
import com.google.common.cache.RemovalNotification
import javafx.scene.image.Image
import java.util.concurrent.TimeUnit


/**
 * the class is for cache icon
 * @author Administrator
 */
object ImageManager {


    val ICON_TEXTURE = "icon_image"
    val ICON_MOVIECLIP = "icon_movieclip"
    val ICON_MUSIC = "icon_music"


    /**
     * cache
     */
    private val map = CacheBuilder.newBuilder()
            .maximumSize(1000)
            .expireAfterAccess(10, TimeUnit.MINUTES)
            .removalListener<String, Image>(this::logReason)
            .build(createLoader({key->_loadImage(key)}))

    private fun<T,V> createLoader(func:(T)->V): CacheLoader<T,V> {

        return  object:CacheLoader<T, V>() {
            @Throws(Exception::class)
            override fun load(key: T): V {
                return func(key)
            }
        }
    }




    private fun logReason(it: RemovalNotification<String, Image>) {
        val str="cause:${it.cause} key:${it.key} value:${it.value}"
        println(str)
        //record to log file

    }


    /**
     * load image by file path
     */
    private fun _loadImage(name: String)= Image((Config.IMAGE_DIRECTORY + name + ".png").toFile().toURL().toString())



    fun getIcon(name: String): Image {
        return map[name]?:throw RuntimeException("$name do not exist!")
    }


}
